src/log-store/src/kafka/index.rs

```diff
@@ -0,0 +1,331 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{max, min};
+use std::collections::VecDeque;
+use std::iter::Peekable;
+use std::marker::PhantomData;
+use std::ops::{Add, Mul, Range, Sub};
+
+use chrono::format::Item;
+use itertools::Itertools;
+use store_api::logstore::EntryId;
+
+use super::util::range::{ConvertIndexToRange, MergeRange};
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct NextBatchHit {
+    pub(crate) bytes: usize,
+    pub(crate) len: usize,
+}
+
+#[cfg(test)]
+impl NextBatchHit {
+    pub fn new(bytes: usize, len: usize) -> NextBatchHit {
+        NextBatchHit { bytes, len }
+    }
+}
+
+/// An iterator over WAL (Write-Ahead Log) entries index for a region.
+pub trait RegionWalIndexIterator: Send + Sync {
+    /// Returns next batch hit.
+    fn next_batch_hit(&self, avg_size: usize, max_gap_size: usize) -> Option<NextBatchHit>;
+
+    // Peeks the next EntryId without advancing the iterator.
+    fn peek(&self) -> Option<EntryId>;
+
+    // Advances the iterator and returns the next EntryId.
+    fn next(&mut self) -> Option<EntryId>;
+}
+
+/// Represents a range [next_entry_id, end_entry_id) of WAL entries for a region.
+pub struct RegionWalRange {
+    next_entry_id: EntryId,
```

Comment:
I think this field should be named as `current_entry_id`, in that `peek` method returns `Some(self.next_entry_id)`.